File: C++/18.中介者模式/MediatorPatternDemo.cpp
/**
 * @file MediatorPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-19
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "User.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<User> robert = std::make_shared<User>("Robert");
    std::shared_ptr<User> john = std::make_shared<User>("John");

    robert->SendMessage("Hi! John!");
    john->SendMessage("Hello! Robert!");

    return 0;
}

// g++ -std=c++11 -g MediatorPatternDemo.cpp ChatRoom.cpp User.cpp
// ./a.out

File: Objective-C/18.MediatorPattern/MediatorPatternDemo.m
//
//  MediatorPatternDemo.m
//  18.MediatorPattern
//
//  Created by Layne on 2020/11/11.
//

#import "User.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        User* robert = [[User alloc] initWithName:@"Robert"];
        User* john = [[User alloc] initWithName:@"John"];
        [robert sendMessage:@"Hi John!"];
        [john sendMessage:@"Hello! Robert!"];
    }
    return 0;
}
